database/src/main/java/com/firebase/ui/database/FirebaseArray.java

```diff
@@ -14,52 +14,137 @@
-        mQuery.removeEventListener((ValueEventListener) this);
-        mQuery.removeEventListener((ChildEventListener) this);
+    private static void checkNotNull(Object o) {
+        if (o == null) throw new IllegalArgumentException("Listener cannot be null.");
     }
 
-    public int getCount() {
-        return mSnapshots.size();
+    /**
+     * Add a listener for change events and errors occurring at the location provided in {@link
+     * #FirebaseArray(Query)}.
+     *
+     * @param listener the listener to be called with changes
+     * @return a reference to the listener provided. Save this to remove the listener later
+     * @throws IllegalArgumentException if the listener is null
+     */
+    public ChangeEventListener addChangeEventListener(@NonNull ChangeEventListener listener) {
+        checkNotNull(listener);
+
+        mListeners.add(listener);
+        notifySubscriptionEventListeners(SubscriptionEventListener.EventType.ADDED);
+        if (mListeners.size() <= 1) { // Only start listening when the first listener is added
+            mQuery.addChildEventListener(this);
+            mQuery.addValueEventListener(this);
+        } else {
+            for (int i = 0; i < size(); i++) {
+                listener.onChildChanged(ChangeEventListener.EventType.ADDED, i, -1);
+            }
+        }
+
+        return listener;
     }
 
-    public DataSnapshot getItem(int index) {
-        return mSnapshots.get(index);
+    /**
+     * Remove a {@link ChangeEventListener} from the location provided in {@link
+     * #FirebaseArray(Query)}. The list will be empty after this call returns.
+     *
+     * @param listener the listener to remove
+     */
+    public void removeChangeEventListener(@NonNull ChangeEventListener listener) {
+        mListeners.remove(listener);
+        notifySubscriptionEventListeners(SubscriptionEventListener.EventType.REMOVED);
+        if (mListeners.isEmpty()) {
+            mQuery.removeEventListener((ValueEventListener) this);
+            mQuery.removeEventListener((ChildEventListener) this);
+            mSnapshots.clear();
+        }
     }
 
-    private int getIndexForKey(String key) {
-        int index = 0;
-        for (DataSnapshot snapshot : mSnapshots) {
-            if (snapshot.getKey().equals(key)) {
-                return index;
-            } else {
-                index++;
+    /**
+     * Add a listener for subscription events eg additions/removals of {@link ChangeEventListener}s.
+     *
+     * @param listener the listener to be called with changes
+     * @return a reference to the listener provided. Save this to remove the listener later
+     * @throws IllegalArgumentException if the listener is null
+     */
+    public SubscriptionEventListener addSubscriptionEventListener(@NonNull SubscriptionEventListener listener) {
+        checkNotNull(listener);
+
+        for (SubscriptionEventListener ignored : mSubscribers) {
+            listener.onSubscriptionAdded();
```

Comment:
So what happens here is when we add a listener we tell it about all existing subscribers, but we don't tell existing subscribers about the new listener?  Again I think I am missing the overall concept because I find this block confusing.